lib/token.cpp

```diff
@@ -1539,6 +1539,30 @@ std::string Token::astStringVerbose() const
     return ret;
 }
 
+void Token::astStringZ3Recursive(std::string& ret) const
+{
+    if (astOperand1())
+        ret += "(";
+    ret += str();
+    if (astOperand1())
+    {
+        ret += " ";
+        astOperand1()->astStringZ3Recursive(ret);
+        if (astOperand2())
+        {
+            ret += " ";
+            astOperand2()->astStringZ3Recursive(ret);
+        }
+        ret += ")";
+    }
+}
+
+std::string Token::astStringZ3() const
```

Comment:
@danmar I was under the impression that you either already had this or already had something that was essentially this. As such I'm assuming this is reinventing the wheel. Please point me to the code I should be calling and I'll get rid of this version.

Reply:
Oh I thought a simple tweak in `testAst` would be enough.

I think this could work:

    std::string Token::astStringZ3() const {
        std::string ret = mStr;
        if (mImpl->mAstOperand1)
            ret += " " + mImpl->mAstOperand1->astStringZ3();
        if (mImpl->mAstOperand2)
            ret += " " + mImpl->mAstOperand2->astStringZ3();
        if (mImpl->mAstOperand1 || mImpl->mAstOperand2)
            ret = "(" + ret + ")";
        return ret;
    }

Reply:
Actually that reminds me to ask: Is it valid for mAstOperand2 to be non-null if mAstOperand1 is null? I had assumed not.

Reply:
any reason to use mImpl->mAstOperand1 rather than astOperand1()?

Reply:
Also, good point. I'd copied astStringVerbose but I don't need to accommodate the additional complexity that the indentation adds.
What are your thoughts on
```
std::string Token::astStringZ3() const {
    if (astOperand1()) {
        if (astOperand2())
            return "(" + str() + " " + astOperand1()->astStringZ3() + " " + astOperand2()->astStringZ3() + ")";
        else
            return "(" + str() + " " + astOperand1()->astStringZ3() + ")";
    }
    else
        return str();
}
```
so you can see the three possible forms clearly enumerated?

Reply:
> Actually that reminds me to ask: Is it valid for mAstOperand2 to be non-null if mAstOperand1 is null? I had assumed not.

that is invalid basically. But I am not sure if I can guarantee that 100% if there is weird non-standard code..
If astOperand2 is set then we can interpret that as a binary operator.

> any reason to use mImpl->mAstOperand1 rather than astOperand1()?

No I guess not. I assume it's inlined so it generates the same code anyway.

Reply:
> What are your thoughts on

that's fine